### DISCUSSION TITLE
FSI model setup with inlet pressure

### DISCUSSION THREAD

**Discussion by lemevel** [idaholab/moose#19011] [Q&A Modules: General] [answered]
Hi all,

I am trying to setup a FSI problem modeling the flow of viscous fluid through a conduit to a fluid-filled cavity and the deformation of the surrounding rocks. This model is described in this paper: https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2016JB013066
(see sketch of Figure 1b)

To start with I am trying to benchmark the case of a constant pressure applied at the base of the conduit. Following this constant inlet pressure I expect the pressure in the chamber to increase until it reaches the inlet pressure (with rates exponentially decreasing - see analytical solution and figures attached). The uplift at the surface should also follow this temporal evolution.

For the MOOSE input file I followed the FSI example called: fsi_flat_channel.i

In the MOOSE results, after the application of the pressure at the inlet, it does not seem to propagate in the conduit and raise the pressure in the chamber. As a consequence there is almost no deformation modeled at the surface. I have an analytical solution that I am comparing the numerical solution with for maximum uplift at the surface and pressure evolution at the exit of the conduit/entrance of the chamber. Both the magnitude and temporal evolution seem off by several orders of magnitude.

I have attached a picture of my mesh, the expected pressure and displacement evolution (analytical) and screenshots of modeled fluid pressure and displacement in MOOSE. My input file is copied below.

Do you have any suggestion for solving this problem?

Thank you,

Hélène

![expected_disp](https://user-images.githubusercontent.com/8866454/135926141-acf64a03-d271-4e3f-b738-0220f2cb6173.png)
![expected_outlet_P](https://user-images.githubusercontent.com/8866454/135926142-be0fea81-9c06-43c7-961f-a971419b632c.png)
<img width="689" alt="Screen Shot 2021-10-04 at 4 52 54 PM" src="https://user-images.githubusercontent.com/8866454/136082174-c8960a7f-d619-4602-a768-00d5c198817f.png">
<img width="599" alt="Screen Shot 2021-10-05 at 2 41 41 PM" src="https://user-images.githubusercontent.com/8866454/136085114-07fd9884-cfa6-452b-889c-c1d2312bc6ca.png">
<img width="595" alt="Screen Shot 2021-10-05 at 2 41 50 PM" src="https://user-images.githubusercontent.com/8866454/136085115-61ed9480-7c63-42b4-aea5-9750673b07b4.png">

```
[GlobalParams]
  gravity = '0 0 0'
  integrate_p_by_parts = true
  laplace = true
  convective_term = true
  transient_term = true
  pspg = true
  displacements = 'disp_x disp_y'
[]

[Problem]
  coord_type = RZ
[]

[Mesh]
  file = mogi_conduit_ns_HLM.msh # sphere/conduit
[]

[Variables]
  [./vel_x]
    block = 'domain_liquid'
  [../]
  [./vel_y]
    block = 'domain_liquid'
  [../]
  [./p]
    block = 'domain_liquid'
  [../]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [./vel_x_solid]
    block = 'domain_solid'
  [../]
  [./vel_y_solid]
    block = 'domain_solid'
  [../]
[]

[Kernels]
  [./vel_x_time]
    type = INSMomentumTimeDerivative
    variable = vel_x
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./vel_y_time]
    type = INSMomentumTimeDerivative
    variable = vel_y
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./mass]
    type = INSMass
    variable = p
    u = vel_x
    v = vel_y
    p = p
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./x_momentum_space]
    type = INSMomentumLaplaceForm
    variable = vel_x
    u = vel_x
    v = vel_y
    p = p
    component = 0
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./y_momentum_space]
    type = INSMomentumLaplaceForm
    variable = vel_y
    u = vel_x
    v = vel_y
    p = p
    component = 1
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./vel_x_mesh]
    type = ConvectedMesh
    disp_x = disp_x
    disp_y = disp_y
    variable = vel_x
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./vel_y_mesh]
    type = ConvectedMesh
    disp_x = disp_x
    disp_y = disp_y
    variable = vel_y
    block = 'domain_liquid'
    use_displaced_mesh = true
  [../]
  [./disp_x_fluid]
    type = Diffusion
    variable = disp_x
    block = 'domain_liquid'
  [../]
  [./disp_y_fluid]
    type = Diffusion
    variable = disp_y
    block = 'domain_liquid'
  [../]
  [./accel_tensor_x]
    type = CoupledTimeDerivative
    variable = disp_x
    v = vel_x_solid
    block = 'domain_solid'
  [../]
  [./accel_tensor_y]
    type = CoupledTimeDerivative
    variable = disp_y
    v = vel_y_solid
    block = 'domain_solid'
  [../]
  [./vxs_time_derivative_term]
    type = CoupledTimeDerivative
    variable = vel_x_solid
    v = disp_x
    block = 'domain_solid'
  [../]
  [./vys_time_derivative_term]
    type = CoupledTimeDerivative
    variable = vel_y_solid
    v = disp_y
    block = 'domain_solid'
  [../]
  [./source_vxs]
    type = MatReaction
    variable = vel_x_solid
    block = 'domain_solid'
    mob_name = 1
  [../]
  [./source_vys]
    type = MatReaction
    variable = vel_y_solid
    block = 'domain_solid'
    mob_name = 1
  [../]
[]

[InterfaceKernels]
  [./penalty_interface_x]
    type = CoupledPenaltyInterfaceDiffusion
    variable = vel_x
    neighbor_var = disp_x
    secondary_coupled_var = vel_x_solid
    boundary = 'sl_interface'
    penalty = 1e6
  [../]
  [./penalty_interface_y]
    type = CoupledPenaltyInterfaceDiffusion
    variable = vel_y
    neighbor_var = disp_y
    secondary_coupled_var = vel_y_solid
    boundary = 'sl_interface'
    penalty = 1e6
  [../]
[]

[Modules/TensorMechanics/Master]
  [./solid_domain]
    strain = SMALL
    incremental = false
    generate_output = 'strain_xx strain_yy strain_zz' ## Not at all necessary, but nice
    block = 'domain_solid'
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2.5e9
    poissons_ratio = 0.25
    block = 'domain_solid'
  [../]
  [./small_stress]
    type = ComputeLinearElasticStress
    block = 'domain_solid'
  [../]
  [./const]
    type = GenericConstantMaterial
    block = 'domain_liquid'
    prop_names = 'rho mu'
    prop_values = '2700  1e5'
  [../]
[]

[BCs]
  [./inlet_p]
    type = DirichletBC
    variable = p
    boundary = 'liquid_inlet'
    value = 1e6
  [../]
  [./no_disp_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'solid_right solid_bottom liquid_inlet'
    value = 0
  [../]
  [./no_disp_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'solid_bottom liquid_inlet solid_right'
    value = 0
  [../]
  [./solid_x_no_slip]
    type = DirichletBC
    variable = vel_x_solid
    boundary = 'solid_right solid_bottom'
    value = 0.0
  [../]
  [./solid_y_no_slip]
    type = DirichletBC
    variable = vel_y_solid
    boundary = 'solid_bottom solid_right'
    value = 0.0
  [../]
  [./fluid_x_no_slip]
    type = DirichletBC
    variable = vel_x
    boundary = 'liquid_inlet liquid_left'
    value = 0.0
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]
 
[Executioner]
 type = Transient
 solve_type = 'PJFNK'
  num_steps = 20
  dt = 5
  dtmin = 1
  
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  line_search = none
[]

[Outputs]
  [./out]
    type = Exodus
  [../]
[]
```

**Comment by GiudGiud** [comment]
Hello

So the idea is that the increased pressure will move the solid boundaries until equilibrium? Do you get a linearly varying pressure profile from just the gravity term currently?

Another difference I see with the example is that the example's domains are more constraints. The no slip / no displacement BCs seem to be completely wrapping around the two phases, except for the fluid inlet. 

@somu15 any ideas?

Guillaume

**Reply by lemevel** [reply]
Hi,
Thanks for your answer. Yes the applied pressure at the inlet should generate incompressible viscous flow in the conduit, which will increase the pressure in the "chamber" and the above solid (rocks) will deform leading to displacement of the free surface. I have attached a sketch of the problem to help visualizing.
I am unsure I understand the test you suggested with the gravity term. When I disable the inlet pressure and enable gravity this way in the input file:
[GlobalParams]
  gravity = '0 9.8 0'
the resulting pressure distribution in the chamber does not make sense (positive in the top half and negative in lower half). I can only get this problem to converge for few very small steps (0.1s).

I agree that the channel example is relatively different since they do not have a free surface and there is an outlet for the fluid.

I suspect there is something wrong in the way I setup the boundary conditions and/or the kernels for this problem but cannot pinpoint which one.

Helene
![FSI_sketch](https://user-images.githubusercontent.com/8866454/137219525-c83c6928-73da-4067-8eac-ab38f17e9a13.jpg)

**Reply by somu15** [reply]
Thanks @GiudGiud. @lemevel I will have to take a look at this more closely. But do you want to model the fluid as an acoustic medium? See this: https://mooseframework.inl.gov/modules/fsi/fsi_acoustics.html

When you use an acoustic fluid, you may use the free surface BC to simulate gravity effects. See this: https://mooseframework.inl.gov/source/bcs/FluidFreeSurfaceBC.html

**Reply by geoyanzhan3** [reply]
Thanks @somu15! The free surface is for solid, not liquid. So I don't think it is an acoustic fluid.

**Reply by lemevel** [reply]
Yes, we are not modeling an acoustic fluid. The purpose of the applied pressure is to simulate a mass injection to the system.

**Reply by somu15** [reply]
Thanks @geoyanzhan3 @lemevel . As I understand, you are applying a transient pressure BC on the fluid at the bottom in the red shaded zone. What is your BC on the left part of the fluid (i.e., red shaded zone)?

**Reply by geoyanzhan3** [reply]
@somu15 The left is an axisymmetric boundary.

**Reply by somu15** [reply]
@geoyanzhan3 @lemevel I thought about this a little bit. Although it seems like you can use an acoustic fluid (which is like a linearized Navier-Stokes with only pressure as the primary variable), the acoustic formulation in MOOSE does not consider fluid viscosity. I don't know what the effect of ignoring viscosity for your problem would be. But I can say that solving the acoustic problem is simpler.

If you want to use INS kernels, @fdkong  and @lindsayad might have some thoughts on this. 

Thank you.

**Reply by lemevel** [reply]
Thanks for these thoughts. However we can not ignore viscosity as this is one of the governing parameter for the viscous flow (similar to Poiseuille flow) and impact the temporal evolution of the flow and thus the temporal evolution of the solid deformation and surface displacement.

**Reply by geoyanzhan3** [reply]
@somu15 Thank you very much! I think viscosity is important in our model.

**Reply by lynnmunday** [reply]
I'm not at all familiar with this module so this is just a guess.  I think the penalty on CoupledPenaltyInterfaceDiffusion should probably be larger for your case.  The original input file used used penalty = 1e6 with a youngs_modulus = 1e2.   Your youngs_modulus=2.5e9 so I think your penalty should be ~1e9 or as large as you can make it and still get the system to converge.

**Reply by jiangwen84** [reply]
@lemevel Can you attach your mesh file?

**Reply by geoyanzhan3** [reply]
Thank you! @jiangwen84. Here is the mesh file
[mogi_conduit_ns_HLM.msh.zip](https://github.com/idaholab/moose/files/7354094/mogi_conduit_ns_HLM.msh.zip)
.

**Reply by jiangwen84** [reply]
One thing I noticed right now is that you want to use `INSMassRZ` and `INSMomentumLaplaceFormRZ` for the RZ problems. We probably should throw a warning if users used wrong kernels for their coord systems.

**Reply by lemevel** [reply]
@jiangwen84 Thank you for the suggestion! Let me replace those kernel types and see if that solves the problem.

**Reply by lemevel** [reply]
@jiangwen84 Changing the kernel types to the RZ ones and increasing the penalty did increase the measured pressure in the chamber and resulting surface displacement but we are still very far (several orders of magnitudes) from the expected analytical solution.

**Comment by fdkong** [comment, marked as answer]
@lemevel 

You might use regular coordinates to make things simple as the first step (also for debugging). We can switch to RZ once the regular coordinates work.  So far, there is at least one thing you might need to improve to make your application work.

```
  [./inlet_p]
    type = DirichletBC
    variable = p
    boundary = 'liquid_inlet'
    value = 1e6
  [../]
```

The pressure inlet might be implemented weakly, and pressure `DirichletBC` might introduce instabilities. You could integrate by part and then replace "p" with the inlet pressure. We do not have this kind of BC yet, but it is easy to implement if you would like to do so.

**Reply by fdkong** [reply]
Alternatively, you could take a look into` modules/navier_stokes/test/tests/ins/pressure_channel/open_bc_pressure_BC.i`

When applying pressure  `DirichletBC`, we do not integrate by part.

@lindsayad knows more on this

**Reply by lemevel** [reply]
@fdkong  Thanks for your answers. We are still stuck with this problem. If I turn off "integrate by parts" and still use the DirichletBC the problem does not converge.
Sorry I am unsure what you suggest here: "You could integrate by part and then replace "p" with the inlet pressure. We do not have this kind of BC yet, but it is easy to implement if you would like to do so."
Would you have an example input file that shows how to implement it?
Thank you.

**Reply by lindsayad** [reply]
So if you take the test that @fdkong referenced and you turn off the outlet pressure boundary condition, shrink the problem, and then run with `-pc_type svd -pc_svd_monitor`, then you'll see that the problem is actually singular. I wouldn't be surprised if you have the same issue here. Can you shrink your problem (to less than 1000 dofs roughly), and then run with `-pc_type svd -pc_svd_monitor` to check whether your problem is singular?

**Reply by geoyanzhan3** [reply]
Thank you @lindsayad @fdkong! We have tested a shrunk problem with ~600 DOF. The message shows "0 of 595 singular values are (nearly) zero". I don't know if that is what you expected. I have attached the shrunk mesh file, moose input file, and the error message for your convenience.

[shrink_singular_test.zip](https://github.com/idaholab/moose/files/7500184/shrink_singular_test.zip)

**Reply by fdkong** [reply]
I think a little bit more about this. We enforce the interface conditions that already take "integrate by parts." And then we can not turn off  "integrate by parts."

The solution here is to implement a weak pressure inlet. 

> Would you have an example input file that shows how to implement it?

You might need to write a C++ code to implement an integrated BC on this.

**Reply by geoyanzhan3** [reply]
After taking a close look at our failed solution, we found the inlet velocity is negative. It means there is backflow in the inlet. We wonder if the Dirichlet pressure BC is not sufficient to define this problem. As the velocity field may not be numerically unique to fulfill the boundary condition. 

We further take a look at the successful solution by COMSOL Multiphysics. In COMSOL, there is an option called "suppress backflow" in the definition of inlet BC. I'm attaching the equations as well. 

Do you think if the backflow matters? @fdkong @lindsayad Thank you very much!
-Yan and @lemevel 

Vertical component of fluid velocity (left bottom corner is the inlet) (failed solution )
![image](https://user-images.githubusercontent.com/35808906/142027758-4cfd5ff4-9de3-43ca-8675-4cd0f57c8896.png)

Problem description
![MagmaInjection](https://user-images.githubusercontent.com/35808906/142031017-93afe1af-1f29-49f9-b1f9-a6d928c93e11.png)

**Reply by lindsayad** [reply]
> We wonder if the Dirichlet pressure BC is not sufficient to define this problem.

This was my concern as well. As I noted earlier if I make no specification of velocity at inlet or outlet and I only specify pressure at either the inlet or outlet of a channel flow test we have, then we have a singular system. But in https://github.com/idaholab/moose/discussions/19011#discussioncomment-1608684 you all mentioned that your system is *not* singular with the current setup. But then @fdkong weighed in and said that the interface kernel assumes that we have integrated the pressure by parts (which is not typically what we would do when we have to specify Dirichlet BCs for pressure).

My most fundamental concern, probably based on my limited understanding of the current problem, is: you have an inlet; do you have an outlet? If not, where does the new mass go?

**Reply by jiangwen84** [reply]
@geoyanzhan3  Do you know how the `suppress backflow` is implemented in COMSOL? A quick check will be to disable the `suppress backflow` in COMSOL, and then compare the results with MOOSE.

**Reply by fdkong** [reply]
> is: you have an inlet; do you have an outlet? If not, where does the new mass go?

I think it is very similar to how we blow a balloon. And there is no outlet. More mass makes the balloon larger.

@geoyanzhan3 If the BC equation (inlet pressure) in the picture is what you want, you might not use Dirichlet BC because it does not seem like a Dirichlet BC. It is a kind of NeumannBC. You could do something like this (need to turn on `integrate by part`). 

```
  [./vec_y_bc]
    type = NeumannBC
    variable = vel_y
    boundary = 'whatever'
    value = whatever_p # might need a negative sign to get the direction right. (from outside to inside)
  [../]
```

You have "u_{fluid} \dot t". What is "t"? is it a tangental vector?

**Reply by geoyanzhan3** [reply]
Thank you @fdkong ! I think you are right: 't' is a tangential vector.

Also, I think besides the NeumannBC, we at least need one Dirichlet BC right? Is $u_{fluid} \dot t = 0$ sufficient to define the question?

Thank you @jiangwen84 ! The equations for suppressing backflow is p^_0 >p_0 as shown in the problem description figure.

**Reply by jiangwen84** [reply]
@geoyanzhan3 I agree with @fdkong that the BC you shown in that figure is a traction BC and it is a Neumann BC type. 

I still did not understand what is what is p^0 and p0 in those equations. What is the value of p0? What do we do if p^0 < p0?
